File: cpp/ballsbot/geometry.cpp
#include <cmath>
#include "geometry.h"

LinearCoefficients GetLinearCoefficients(const Point p1, const Point p2) {
    LinearCoefficients result = {p1.y - p2.y, p2.x - p1.x, p1.x * p2.y - p2.x * p1.y};
    return result;
}

double PointToLineDistance(Point p0, LinearCoefficients a_line) {
    return std::abs(a_line.a * p0.x + a_line.b * p0.y + a_line.c) /
           sqrt(a_line.a * a_line.a + a_line.b * a_line.b);
}

double Distance(Point one, Point two) {
    return sqrt((one.x - two.x) * (one.x - two.x) + (one.y - two.y) * (one.y - two.y));
}

LinearCoefficients NormalToLineInPoint(LinearCoefficients a_line, Point a_point) {
    if (a_line.b == 0.) {
        LinearCoefficients result = {0., 1., -a_point.y};
        return result;
    } else {
        LinearCoefficients result = {1., -a_line.a / a_line.b,
                                     -a_point.x + a_line.a / a_line.b * a_point.y};
        return result;
    }
}

std::pair<LinearCoefficients, LinearCoefficients> GetTwoNRadiansLines(double angle,
                                                                      LinearCoefficients a_line,
                                                                      Point a_point) {
    std::pair<LinearCoefficients, LinearCoefficients> result;

    result.first.b = 1.;
    result.second.b = 1.;

    if (a_line.a == 0.) {
        result.first.a = tan(angle);
        result.second.a = tan(-angle);
    } else if (a_line.b == 0.) {
        result.first.a = tan(M_PI / 2. - angle);
        result.second.a = tan(M_PI / 2. + angle);
    } else {
        result.first.a = tan(atan(a_line.a / a_line.b) + angle);
        result.second.a = tan(atan(a_line.a / a_line.b) - angle);
    }

    result.first.c = -result.first.a * a_point.x - result.first.b * a_point.y;
    result.second.c = -result.second.a * a_point.x - result.second.b * a_point.y;

    return result;
}

bool OnOtherSide(LinearCoefficients a_line, Point p1, Point p2) {
    auto second_line = GetLinearCoefficients(p1, p2);
    Point p0;

    if (a_line.b == 0. and second_line.b == 0.) {
        return false;
    } else if (a_line.b == 0.) {
        p0.x = -a_line.c / a_line.a;
        p0.y = -(second_line.a * p0.x + second_line.c) / second_line.b;
    } else if (second_line.b == 0.) {
        p0.x = -second_line.c / second_line.a;
        p0.y = -(a_line.a * p0.x + a_line.c) / a_line.b;
    } else {
        double nom = a_line.c / a_line.b - second_line.c / second_line.b;
        double denom = second_line.a / second_line.b - a_line.a / a_line.b;
        if (denom == 0.) {
            return false;
        } else {
            p0.x = nom / denom;
            p0.y = -(a_line.a * p0.x + a_line.c) / a_line.b;
        }
    }

    if (p1.x > p2.x) {
        if (p1.x <= p0.x || p0.x <= p2.x) {
            return false;
        }
    } else if (p1.x < p2.x) {
        if (p1.x >= p0.x || p0.x >= p2.x) {
            return false;
        }
    }

    if (p1.y > p2.y) {
        if (p1.y <= p0.y || p0.y <= p2.y) {
            return false;
        }
    } else if (p1.y < p2.y) {
        if (p1.y >= p0.y || p0.y >= p2.y) {
            return false;
        }
    }

    return true;
}

File: python/lib/ballsbot/geometry.py
from math import sqrt, tan, atan, pi
from ballsbot_routing import ballsbot_routing


def get_linear_coefs(p1, p2):
    a = p1[1] - p2[1]
    b = p2[0] - p1[0]
    c = p1[0] * p2[1] - p2[0] * p1[1]
    return a, b, c


def point_to_line_distance(p0, a, b, c):
    return abs(a * p0[0] + b * p0[1] + c) / sqrt(a * a + b * b)


distance = ballsbot_routing.distance


def normal_to_line_in_point(line_coefs, a_point):
    a, b, _ = line_coefs
    x0, y0 = a_point
    if b == 0.:
        return 0., 1., -y0
    else:
        b2_d_a2 = -a / b
        c2_d_a2 = -x0 + a / b * y0
        return 1., b2_d_a2, c2_d_a2


def get_two_n_radians_lines(angle, line_coefs, a_point):
    a, b, _ = line_coefs
    x0, y0 = a_point
    if a == 0.:
        a1, b1 = (tan(angle), 1.)
        a2, b2 = (tan(-angle), 1.)
    elif b == 0:
        a1, b1 = (tan(pi / 2 - angle), 1.)
        a2, b2 = (tan(pi / 2 + angle), 1.)
    else:
        a1 = tan(atan(a / b) + angle)
        b1 = 1.
        a2 = tan(atan(a / b) - angle)
        b2 = 1.
    c1 = -a1 * x0 - b1 * y0
    c2 = -a2 * x0 - b2 * y0

    return (a1, b1, c1), (a2, b2, c2)


def on_other_side(line1_coefs, p1, p2):  # pylint: disable=R0914
    a1, b1, c1 = line1_coefs
    a2, b2, c2 = get_linear_coefs(p1, p2)
    if b1 == 0. and b2 == 0.:
        return False
    elif b1 == 0.:
        x0 = -c1 / a1
        y0 = -(a2 * x0 + c2) / b2
    elif b2 == 0.:
        x0 = -c2 / a2
        y0 = -(a1 * x0 + c1) / b1
    else:
        nom = c1 / b1 - c2 / b2
        denom = a2 / b2 - a1 / b1
        if denom == 0.:
            return False
        else:
            x0 = nom / denom
            y0 = -(a1 * x0 + c1) / b1

    x1, y1 = p1
    x2, y2 = p2

    if x1 > x2:
        if not x1 > x0 > x2:
            return False
    else:
        if not x1 < x0 < x2:
            return False

    if y1 > y2:
        if not y1 > y0 > y2:
            return False
    else:
        if not y1 < y0 < y2:
            return False

    return True


def get_radial_line(center_point, angle):
    a, b = (tan(angle), 1.)
    c = -a * center_point[0] - b * center_point[1]
    return a, b, c


def cross_point(first_line, second_line):
    a1, b1, c1 = first_line
    a2, b2, c2 = second_line

    if b1 == 0. and b2 == 0.:
        raise NotImplementedError()
    elif b1 == 0.:
        x = - c1 / a1
        y = -a2 / b2 * x - c2 / b2
    elif b2 == 0.:
        x = - c2 / a2
        y = -a1 / b1 * x - c1 / b1
    else:
        x = (c1 / b1 - c2 / b2) / (a2 / b2 - a1 / b1)
        y = -a1 / b1 * x - c1 / b1
    return x, y
